File: Priority_Queue/502. IPO/502. IPO.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        vector<pair<int, int>>tasks;
        int n = profits.size();
        for (int i = 0; i < n; i++) {
            tasks.push_back({capital[i], profits[i]});
        }
        sort(tasks.begin(), tasks.end());
        
        priority_queue<int>pq;
        int rets = w;
        int count = 0;
        int idx = 0;
        while (count < k) {
            while (idx < n && tasks[idx].first <= rets) {
                pq.push(tasks[idx].second);
                idx++;
            }
            if (pq.empty())  break;
            rets += pq.top();
            pq.pop();
            count++;
        }
        return rets;
    }
};

File: Priority_Queue/502. IPO/502. IPO.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Integer> freepq = new PriorityQueue<>((a, b) -> (b-a));
        int n = profits.length;
        int[][] busy = new int[n][2];

        for (int i = 0; i < profits.length; i++) {
            busy[i][0] = capital[i];
            busy[i][1] = profits[i];
        }
        Arrays.sort(busy, (a,b) ->(a[0]-b[0]));

        int rets = w;
        int loop = Math.min(k, n);
        int idx = 0;
        for (int i = 0; i < loop; i++) {
            while (idx < n && busy[idx][0] <= rets) {
                freepq.offer(busy[idx][1]);
                idx++;
            }
            if (freepq.isEmpty())  break;
            rets += freepq.poll();
        }
        return rets;
    }
}
// [w, c]
// freeBQ
// busyBQ
